Validation warnings

featalion opened the pull request:
Print configuration file validation warnings if:
- no files to watch supplied
- port is not defined, when host is

Exit with error code if no files to watch specified at all (in config and via command line).

Related to #43

featalion committed:
adds validation warning to config. exit with error when supplied no files to watch

troy commented on config_manager.go:
```diff
@@ -35,6 +35,16 @@ type ConfigFile struct {
 
+func (cf ConfigFile) printValidationWarnings() {
+	if len(cf.Files) == 0 {
+		log.Warningf("Configuration file contains no files to watch")
```
@leonsodhi: Do you think this is worth printing?

leonsodhi replied:
I'd say the more error handling the better. Silence is a right PITA to troubleshoot, and means we have no error message to work from when customers get in touch.

troy replied:
Works for me

troy commented:
@leonsodhi: Code looks good to me. Can you edit the error messages (and if you think they can be clearer, comments), either by merging and creating a new branch+PR or by branching from this branch?

featalion committed:
update VividCortex/godaemon dependency

featalion committed:
print validation warnings before daemonization; add more warnings to configuration manager

featalion commented:
@leonsodhi @troy I added more validation warnings. They will come before daemonization on parent stage or if no daemonization required. To determine the stage, I updated `VividCortex/godaemon` dependency. This update is workaround for #23 .

troy commented:
@leonsodhi this looks good to me other than wording and grammar on the error strings and error code variable name ("Exists"). If it saves you time to put revisions in here for Yury to implement, go for it.

featalion committed:
change var name

leonsodhi commented on config_manager.go:
```diff
@@ -189,7 +206,7 @@ func (cm *ConfigManager) loadConfigFile() error {
 	// don't error if the default config file isn't found
 	if os.IsNotExist(err) && cm.Flags.ConfigFile == DefaultConfigFile {
-		return nil
+		return defaultConfigDoesNotExist
```
It's totally fine for someone to only pass in params via the CLI switches and not have a config file. I'd vote for removing this warning.

featalion replied:
If user misplaced config file, there is no way to know that. In common, users do not read the code to understand why utility fails. They write to support. One notification line can save your and users' time.

leonsodhi replied:
Completely agree on users not wanting to read code on failures, and they definitely shouldn't have to. However, IMO, we should validate that all required parameters have been supplied and not worry about where they came from. If a user specifies all required parameters via the CLI, a config file isn't necessary and thus we can't justify suggesting something is wrong.

The only situation I can think of where this falls down is when an optional parameter is specified in a config file and r_s2 is never asked to load it. One solution to that might be printing the state of all settings when `-D` is used. I'd prefer to hold off doing that ATM as it hasn't been something that comes up in practice. We can always revisit if things change.

Let me know if I've missed a scenario here or there's a hole in my logic.

leonsodhi commented on config_manager.go:
```diff
@@ -35,6 +37,16 @@ type ConfigFile struct {
+	}
+	// Warn, that port is not set, only if host is specified
+	if cf.Destination.Host != "" && cf.Destination.Port == 0 {
+		log.Warningf("Validation: destination port is not specified in configuration file")
```
@troy as we're not defaulting the destination host to `logs.papertrailapp.com`, should we also remove the default port and fail hard if it's not specified?

troy replied:
I think we should leave the default port (and anything else that currently has a default value). A possible improvement would need to be very meaningful to justify breaking installations which haven't set a port.


On Jun 29, 2015, at 05:12, Leon Sodhi <notifications@github.com> wrote:
> 
> In config_manager.go:
> 
> > @@ -35,6 +37,16 @@ type ConfigFile struct {
> >  	ExcludePatterns *RegexCollection `yaml:"exclude_patterns"`
> >  }
> >  
> > +func (cf ConfigFile) printValidationWarnings() {
> > +	if len(cf.Files) == 0 {
> > +		log.Warningf("Validation: configuration file contains no files to watch")
> > +	}
> > +	// Warn, that port is not set, only if host is specified
> > +	if cf.Destination.Host != "" && cf.Destination.Port == 0 {
> > +		log.Warningf("Validation: destination port is not specified in configuration file")
> @troy as we're not defaulting the destination host to logs.papertrailapp.com, should we also remove the default port and fail hard if it's not specified?
> 
> —
> Reply to this email directly or view it on GitHub.
>

leonsodhi replied:
Also, I am not sure the port could ever end up being zero given the behaviour in [DestPort()](https://github.com/featalion/remote_syslog2/blob/fix-issue-43/config_manager.go#L265), which seems to default to `514`. 

Is the above check necessary?

featalion replied:
`DestPort()` uses magic number, there is no way to understand 514 is default port, except look right into the code. Also, this function only related to configuration file validation and knows nothing about command line arguments.

leonsodhi replied:
@troy It definitely bothers me that someone could accidentally miss the port off and we wouldn't tell them, but I agree that breaking existing configs sucks and has to take priority.

@featalion Sorry, I should have clarified that `514` is assigned to `DefaultPort` (which is a constant) in the same file as the code above. `DestPort()` seems to consider both the CLI and config file settings:

```
case cm.Flags.DestPort != 0: // CLI
    return cm.Flags.DestPort
case cm.Config.Destination.Port != 0: // Config file
    return cm.Config.Destination.Port 
```

Let me know if I am reading this wrong.

featalion replied:
It is constant since 7c67e297d42270fda2e3aa8bcd290e510288a8a8 ( ;

leonsodhi commented on remote_syslog.go:
```diff
@@ -53,6 +54,11 @@ func tailOne(file string, excludePatterns []*regexp.Regexp, logger *syslog.Logge
 // Tails files speficied in the globs and re-evaluates the globs
 // at the specified interval
 func tailFiles(globs []string, excludedFiles []*regexp.Regexp, excludePatterns []*regexp.Regexp, interval RefreshInterval, logger *syslog.Logger, severity syslog.Priority, facility syslog.Priority, poll bool) {
+	if len(globs) == 0 {
```
Can this ever evaluate to `true` given that we're checking [here](https://github.com/featalion/remote_syslog2/blob/fix-issue-43/config_manager.go#L41) and this is invoked via the call made [early on in main](https://github.com/featalion/remote_syslog2/blob/fix-issue-43/remote_syslog.go#L113).

Also, could this message appear on the console when r_s2 is damonized?

leonsodhi replied:
I've just answered my own question: yes and no :-) If we check for files earlier on via `Files()`, that should cover files from both the config & CLI, and will allow us to display an error even when damonized. That should make this redundant and allow it to be removed.

wagerlabs replied:
I actually propose merging config and command-line arguments early in the
process. I can't think of anything that requires keeping both.
On Mon, Jun 29, 2015 at 2:45 PM Leon Sodhi <notifications@github.com> wrote:

> In remote_syslog.go
> <https://github.com/papertrail/remote_syslog2/pull/73#discussion_r33459382>
> :
>
> > @@ -53,6 +54,11 @@ func tailOne(file string, excludePatterns []*regexp.Regexp, logger *syslog.Logge
> >  // Tails files speficied in the globs and re-evaluates the globs
> >  // at the specified interval
> >  func tailFiles(globs []string, excludedFiles []*regexp.Regexp, excludePatterns []*regexp.Regexp, interval RefreshInterval, logger *syslog.Logger, severity syslog.Priority, facility syslog.Priority, poll bool) {
> > +	if len(globs) == 0 {
>
> I've just answered my own question: yes and no :-) If we check for files
> earlier on via Files(), that should cover files from both the config &
> CLI, and will allow us to display an error even when damonized. That should
> make this redundant and allow it to be removed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/papertrail/remote_syslog2/pull/73/files#r33459382>.
>

featalion replied:
It prevents running of remote syslog if no files specified. As r_s2 does not support `-HUP` signal to reload configuration, it has no sense to keep it running with empty files list to watch, imho.

leonsodhi replied:
Agreed that we want to error and exit if no files have been specified. My suggestion was to run the check earlier so it can be printed when daemonized. 

I agree with @wagerlabs that it's likely worth merging everything early on.

leonsodhi commented on config_manager.go:
```diff
@@ -35,6 +37,16 @@ type ConfigFile struct {
 }
 
+func (cf ConfigFile) printValidationWarnings() {
+	if len(cf.Files) == 0 {
```
This only checks for files passed in via the config. We'll likely want to check the output from [Files()](https://github.com/featalion/remote_syslog2/blob/fix-issue-43/config_manager.go#L305) instead so we know if r_s2 isn't monitoring anything.

featalion replied:
This is for the case when config could be malformed, see #43

leonsodhi commented on config_manager.go:
```diff
@@ -353,3 +371,12 @@ func (cm *ConfigManager) ExcludeFiles() []*regexp.Regexp {
+	cm.Config.printValidationWarnings()
+	for _, fileToWatch := range cm.Files() {
+		if _, err := os.Stat(fileToWatch); os.IsNotExist(err) {
+			log.Warningf("Validation: file does not exist: %s", fileToWatch)
```
This warning duplicates [this](https://github.com/featalion/remote_syslog2/blob/fix-issue-43/remote_syslog.go#L82) error when running in the foreground. Is there a situation that the above code wouldn't catch that the linked error would?

As far as I can tell, the difference is that the linked error is checking each file continuously as r_s2 runs. However, moving a file, deleting it, re-pointing a symlink to a non-existent file, etc, may generate file notification events that we can handle separately, so checking for existence once on startup might be enough.

featalion replied:
In the daemon mode there is no log lines in a console when files are absent.

leonsodhi replied:
Good point! Not sure how I missed that. Although, it still seems less than ideal having duplicate errors when running with `-D`. Maybe we don't need to print this when daemonized as the file could appear later on. Leave it with me and I'll think on it.

featalion committed:
constantise magic number (default port)

featalion commented:
Summing all comments above, do you want to have `-D` option for validations warnings? Defaults to false.
